### PR TITLE
#25907 StorageBlobResource performance improvement

### LINKED ISSUE #25907: [BUG] BlobStorageResource seem to require excessive amount of REST API calls
**Describe the bug**
BlobStorageResource seems to be using excessive amount of `exists` calls. Like here:
https://github.com/Azure/azure-sdk-for-java/blob/808ef6761f969295da14a3b4780462ae32a11b8d/sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/storage/resource/BlobStorageResource.java#L85-L105
or
here:
https://github.com/Azure/azure-sdk-for-java/blob/808ef6761f969295da14a3b4780462ae32a11b8d/sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/storage/resource/BlobStorageResource.java#L108-L110
or
here:
https://github.com/Azure/azure-sdk-for-java/blob/808ef6761f969295da14a3b4780462ae32a11b8d/sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/storage/resource/BlobStorageResource.java#L177-L183

Every such call incurs REST API call that results in higher latency and monetary cost than what might be necessary.

Consider using conditional headers and error handling instead of checking for storage resource existence to reduce amount of transactions required to accomplish this functionality.

Related issue in Storage SDK: https://github.com/Azure/azure-sdk-for-java/issues/23658

### PR DESCRIPTION
**Pull request by stliu**
This PR **partially** fixes #25907 ( File share is not fixed yet)
and missing test cases

### CODE REVIEW

**Comment by stliu** [comment]
@saragluna I took sometime on this PR, it is not finished yet, would you please find someone to continue, thanks

**Comment by stliu** [comment]
@kasobol-msft could you take a look of this PR, thanks

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageBlobResource.java [hunk 1]
```diff
@@ -218,38 +215,22 @@ public String getDescription() {
...
     @Override
     public InputStream getInputStream() throws IOException {
         try {
-            assertExisted();
             return this.blockBlobClient.openInputStream();
         } catch (BlobStorageException e) {
-            throw new IOException(MSG_FAIL_OPEN_INPUT, e);
+            if (e.getErrorCode() == BlobErrorCode.CONTAINER_NOT_FOUND
+                || e.getErrorCode() == BlobErrorCode.BLOB_NOT_FOUND) {
+                throw new FileNotFoundException("Blob or container not existed.");
```
Comment (kasobol-msft) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageBlobResource.java:225]:
```suggestion
                throw new FileNotFoundException("Blob or container does not exist.");
```

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java [hunk 2]
```diff
@@ -176,19 +174,17 @@ public String getDescription() {
...
     @Override
     public InputStream getInputStream() throws IOException {
         try {
-            assertExisted();
             return this.shareFileClient.openInputStream();
         } catch (ShareStorageException e) {
-            throw new IOException("Failed to open input stream of cloud file", e);
+            if (e.getErrorCode() == ShareErrorCode.SHARE_NOT_FOUND
+                || e.getErrorCode() == ShareErrorCode.RESOURCE_NOT_FOUND) {
+                throw new FileNotFoundException("Share or file not existed");
```
Comment (kasobol-msft) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:184]:
```suggestion
                throw new FileNotFoundException("Share or file does not exist.");
```

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java [hunk 3]
```diff
@@ -197,24 +193,16 @@ StorageType getStorageType() {
...
             ShareFileHttpHeaders header = null;
             if (StringUtils.hasText(contentType)) {
                 header = new ShareFileHttpHeaders();
                 header.setContentType(contentType);
             }
-            //TODO: create method must provide file length, but we don't know actual
-            this.shareFileClient.createWithResponse(1024, header, null, null, null, null, Context.NONE).getValue();
+            // TODO: create method must provide maximum size, but we don't know actual
+            this.shareFileClient.createWithResponse(1024, header, null, null, null, null, Context.NONE)
+                .getValue();
```
Comment (kasobol-msft) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:205]: Is there a way to make `1024` configurable? This is a pretty small number for average application.

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java [hunk 4]
```diff
@@ -197,24 +193,16 @@ StorageType getStorageType() {
...
-            throw new FileNotFoundException("File or container not existed.");
-        }
-    }
-
     private void create() throws ShareStorageException {
-        if (!shareClient.exists()) {
-            this.shareClient.create();
-        }
-        if (!shareFileClient.exists()) {
+        if (!this.exists()) {
```
Comment (kasobol-msft) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:197]: @rickle-msft @ibrahimrabab Should we introduce `shareFileClient.createIfNotExists(size)` ?

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java (hunk 3 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:205]: Thanks @kasobol-msft, we really should not use this `1024` value here. But in a `Resource` implementation, it's not reasonable for users to define a `maximum size` for a file (if the file does not exist). The users locate a resource using `azure-file://some-share/some-file`. 

So we have to figure out how to make it configurable. How about implementing this feature request in another PR?

Comment (kasobol-msft) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:205]: sounds good. please create GH issue to track.

**Review by kasobol-msft — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java (hunk 3 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/resource/StorageFileResource.java:205]: https://github.com/Azure/azure-sdk-for-java/issues/30120

**Pull request merged**